File: python/backuptool/procs.py
# SPDX-License-Identifier: GPL-3.0-or-later
"""Detect (and optionally close) running applications that commonly hold user
files open, so a file-level backup doesn't fail with "in use" errors.
Cross-platform: tasklist on Windows, ps elsewhere.
"""
from __future__ import annotations

import subprocess
import sys

# (process token, friendly name)
BLOCKERS = [
    ("firefox", "Firefox"), ("chrome", "Google Chrome"), ("msedge", "Microsoft Edge"),
    ("chromium", "Chromium"), ("brave", "Brave"), ("opera", "Opera"),
    ("vivaldi", "Vivaldi"), ("safari", "Safari"), ("thunderbird", "Thunderbird"),
    ("photoshop", "Photoshop"), ("illustrator", "Illustrator"), ("indesign", "InDesign"),
    ("lightroom", "Lightroom"), ("figma", "Figma"), ("acrobat", "Acrobat"),
    ("acrord32", "Acrobat Reader"), ("winword", "Word"), ("excel", "Excel"),
    ("powerpnt", "PowerPoint"), ("outlook", "Outlook"), ("onenote", "OneNote"),
    ("onedrive", "OneDrive"), ("dropbox", "Dropbox"), ("spotify", "Spotify"),
]


def _process_list_lower() -> str:
    try:
        if sys.platform.startswith("win"):
            out = subprocess.run(["tasklist"], capture_output=True, text=True, timeout=30)
        else:
            out = subprocess.run(["ps", "-A", "-o", "comm="], capture_output=True, text=True, timeout=30)
        return out.stdout.lower()
    except OSError:
        return ""


def _token_running(listing: str, token: str) -> bool:
    if sys.platform.startswith("win"):
        return (token + ".exe") in listing
    return token in listing


def running_blockers():
    """Friendly names of running apps likely to lock files (deduplicated)."""
    listing = _process_list_lower()
    found = []
    for tok, name in BLOCKERS:
        if _token_running(listing, tok) and name not in found:
            found.append(name)
    return found


def close_apps():
    """Ask each running blocker to close (graceful). Returns the names signalled."""
    listing = _process_list_lower()
    closed = []
    for tok, name in BLOCKERS:
        if not _token_running(listing, tok):
            continue
        try:
            if sys.platform.startswith("win"):
                subprocess.run(["taskkill", "/IM", tok + ".exe"], capture_output=True, text=True)
            else:
                subprocess.run(["pkill", "-i", tok], capture_output=True, text=True)
        except OSError:
            pass
        if name not in closed:
            closed.append(name)
    return closed

File: rust/src/procs.rs
// SPDX-License-Identifier: GPL-3.0-or-later
//! Detect (and optionally close) running applications that commonly hold user
//! files open, so a file-level backup doesn't fail with "in use" errors.
//! Cross-platform: tasklist on Windows, ps elsewhere.

/// (process token, friendly name). The token is matched against the running
/// process list (as "<token>.exe" on Windows, as a substring elsewhere).
const BLOCKERS: &[(&str, &str)] = &[
    ("firefox", "Firefox"),
    ("chrome", "Google Chrome"),
    ("msedge", "Microsoft Edge"),
    ("chromium", "Chromium"),
    ("brave", "Brave"),
    ("opera", "Opera"),
    ("vivaldi", "Vivaldi"),
    ("safari", "Safari"),
    ("thunderbird", "Thunderbird"),
    ("photoshop", "Photoshop"),
    ("illustrator", "Illustrator"),
    ("indesign", "InDesign"),
    ("lightroom", "Lightroom"),
    ("figma", "Figma"),
    ("acrobat", "Acrobat"),
    ("acrord32", "Acrobat Reader"),
    ("winword", "Word"),
    ("excel", "Excel"),
    ("powerpnt", "PowerPoint"),
    ("outlook", "Outlook"),
    ("onenote", "OneNote"),
    ("onedrive", "OneDrive"),
    ("dropbox", "Dropbox"),
    ("spotify", "Spotify"),
];

fn process_list_lower() -> String {
    #[cfg(windows)]
    let cmd = std::process::Command::new("tasklist").output();
    #[cfg(not(windows))]
    let cmd = std::process::Command::new("ps").args(["-A", "-o", "comm="]).output();
    cmd.ok()
        .map(|o| String::from_utf8_lossy(&o.stdout).to_lowercase())
        .unwrap_or_default()
}

fn token_running(list: &str, token: &str) -> bool {
    #[cfg(windows)]
    {
        list.contains(&format!("{token}.exe"))
    }
    #[cfg(not(windows))]
    {
        list.contains(token)
    }
}

/// Friendly names of running apps likely to lock files (deduplicated).
pub fn running_blockers() -> Vec<String> {
    let list = process_list_lower();
    let mut found: Vec<String> = Vec::new();
    for (tok, name) in BLOCKERS {
        if token_running(&list, tok) && !found.iter().any(|n| n == name) {
            found.push((*name).to_string());
        }
    }
    found
}

/// Ask each running blocker to close (graceful: lets apps prompt to save).
/// Returns the friendly names it signalled. Re-check with [`running_blockers`].
pub fn close_apps() -> Vec<String> {
    let list = process_list_lower();
    let mut closed: Vec<String> = Vec::new();
    for (tok, name) in BLOCKERS {
        if !token_running(&list, tok) {
            continue;
        }
        #[cfg(windows)]
        {
            // graceful close request (no /F) so unsaved work can be saved
            let _ = std::process::Command::new("taskkill")
                .args(["/IM", &format!("{tok}.exe")])
                .output();
        }
        #[cfg(not(windows))]
        {
            let _ = std::process::Command::new("pkill").args(["-i", tok]).output();
        }
        if !closed.iter().any(|n| n == name) {
            closed.push((*name).to_string());
        }
    }
    closed
}

#[cfg(test)]
mod tests {
    #[test]
    fn running_blockers_does_not_panic() {
        // returns a (possibly empty) list without crashing
        let _ = super::running_blockers();
    }
}
